drivers/periph_common/dac.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2020 Beuth Hochschule für Technik Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_dac
+ * @{
+ *
+ * @file
+ * @brief       Common DAC function fallback implementations
+ *
+ * @author      Benjamin Valentin <benpicco@beuth-hochschule.de>
+ *
+ * @}
+ */
+
+#include "board.h"
+#include "kernel_defines.h"
+#include "periph/dac.h"
+#include "periph/timer.h"
+
+#if !defined(PERIPH_DAC_PROVIDES_PLAY) && IS_ACTIVE(MODULE_PERIPH_DAC)
+
+#ifndef DAC0_TIMER
+#define DAC0_TIMER   (TIMER_NUMOF - 1)
+#endif
+#ifndef DAC1_TIMER
+#define DAC1_TIMER   (TIMER_NUMOF - 2)
+#endif
+
+static const tim_t _dac_timer[DAC_NUMOF] = {
+#if DAC_NUMOF > 0
+    DAC0_TIMER,
+#endif
+#if DAC_NUMOF > 1
+    DAC1_TIMER,
+#endif
+};
+
+static struct dac_ctx {
+    const uint8_t *buffers[2];
+    size_t buffer_len[2];
+    size_t idx;
+    uint8_t cur;
+    uint8_t playing;
+} _ctx[DAC_NUMOF];
+
+static void _timer_cb(void *arg, int chan)
+{
+    const dac_cfg_t *cfg = arg;
+    const dac_t dac      = cfg->line;
+    struct dac_ctx *ctx  = &_ctx[dac];
+
+    const uint8_t cur  = ctx->cur;
+    const uint8_t *buf = ctx->buffers[cur];
+    const size_t len   = ctx->buffer_len[cur];
+
+    if (cfg->flags & DAC_FLAG_16BIT) {
+        size_t idx_real = 2 * ctx->idx;
+        dac_set(0, (buf[idx_real + 1] << 8) | buf[idx_real]);
+    } else {
+        dac_set(0, buf[ctx->idx] << 8);
+    }
+
+    if (++ctx->idx >= len) {
+
+        /* invalidate old buffer */
+        ctx->buffer_len[cur] = 0;
+
+        ctx->idx = 0;
+        ctx->cur = !cur;
+
+        if (ctx->buffer_len[!cur] == 0) {
+            ctx->playing = 0;
+            timer_clear(_dac_timer[dac], chan);
+        } else if (cfg->cb) {
+            cfg->cb(cfg->cb_arg);
+        }
+    }
+}
+
+void dac_play(const void *buf, size_t len, const dac_cfg_t *cfg)
+{
+    const unsigned timer_freq  = 1000000;
+    const unsigned sample_rate = cfg->sample_rate * 6; /* XXX why the fudge factor? */
```

Comment:
Wouldn't it work more reliable to internally work with a constant sampling rate? You could convert internally by either skipping over data or or playing samples multiple times. E.g. add a `step` variable, and use `buf[(pos += step) >> 8]` to access the sample.

Putting a constant load on the system regardless of sample frequency would it make easy to create effects like playing the same sample with different pitches, as you could just specify a higher sampling frequency.

Reply:
Sorry for the long delay, I didn't find the time to get back to this PR.

I think that's a good idea, that simplifies the API quite a bit, I guess I wanted to avoid introducing functions to the API, but it's clearer to split it into setup and play than to shove everything into one function.

Reply:
I also figured out the fudge factor - turns out you have to set the 'project sample rate' in Audacity, otherwise re-sampling the tracks will be undone on export :sweat_smile: 
So the sample was still 44.1kHz, that's why it had to be played faster.